src/Stratis.Bitcoin.Features.Consensus/ProvenBlockHeaders/ProvenBlockHeaderRepository.cs

```diff
@@ -0,0 +1,378 @@
+        /// <param name="folder"><see cref="ProvenBlockHeaderStore"/> folder path to the DBreeze database files.</param>
+        /// <param name="dateTimeProvider">Provider of time functions.</param>
+        /// <param name="loggerFactory">Factory to create a logger for this type.</param>
+        public ProvenBlockHeaderRepository(Network network, string folder, IDateTimeProvider dateTimeProvider, ILoggerFactory loggerFactory)
+        {
+            Guard.NotNull(network, nameof(network));
+            Guard.NotNull(folder, nameof(folder));
+
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+
+            // Create the ProvenBlockHeaderStore if it doesn't exist.
+            Directory.CreateDirectory(folder);
+
+            this.dbreeze = new DBreezeEngine(folder);
+            this.network = network;
+        }
+
+        /// <inheritdoc />
+        public Task InitializeAsync()
+        {
+            Task task = Task.Run(() =>
+            {
+                this.logger.LogInformation("Initializing {0}.", nameof(ProvenBlockHeaderRepository));
+
+                Block genesis = this.network.GetGenesis();
+
+                using (DBreeze.Transactions.Transaction transaction = this.dbreeze.GetTransaction())
+                {
+                    if (this.GetTipHashHeight(transaction) == null)
+                    {
+                        // set to genesis
+                        this.SetTip(transaction, new HashHeightPair(genesis.GetHash(), 0));
+
+                        transaction.Commit();
+                    }
+                }
+
+                this.logger.LogTrace("(-)");
+            });
+
+            return task;
+        }
+
+        public Task<Dictionary<int, ProvenBlockHeader>> GetAsync(int fromBlockHeight, int toBlockHeight)
+        {
+            Task<Dictionary<int, ProvenBlockHeader>> task = Task.Run(() =>
+            {
+                this.logger.LogTrace("({0}:'{1}')", nameof(fromBlockHeight), fromBlockHeight);
+                this.logger.LogTrace("({0}:'{1}')", nameof(toBlockHeight), toBlockHeight);
+
+                using (DBreeze.Transactions.Transaction transaction = this.dbreeze.GetTransaction())
+                {
+                    var items = new Dictionary<int, ProvenBlockHeader>();
+
+                    transaction.SynchronizeTables(ProvenBlockHeaderTable);
+
+                    transaction.ValuesLazyLoadingIsOn = false;
+
+                    this.logger.LogTrace("Loading ProvenBlockHeaders from block height '{0}' to '{1}  from the database.",
+                        fromBlockHeight, toBlockHeight);
+
+                    for (int i = fromBlockHeight; i <= toBlockHeight; i++)
+                    {
+                        Row<byte[], ProvenBlockHeader> row =
+                            transaction.Select<byte[], ProvenBlockHeader>(ProvenBlockHeaderTable, i.ToBytes(false));
+
+                        if (row.Exists)
```

Comment:
if it doesn't exist it should throw or insert a null in the collection.

Reply:
ok